### ISSUE TITLE
Support Pick and other Mapped Types

### ISSUE THREAD

**Issue by robertdempsey** [vega/ts-json-schema-generator#24]
Hi! Great library here. We are heavily experimenting with using json schemas generated from our TypeScript interfaces on both front end and back end and this is obviously the glue that's holding it all together! With that said, there are some scenarios we are encountering that we are having to work around to get our schemas generated the way we want to, the main issues being around Mapped Types.

Consider the following interfaces:

```
export interface ISortableProperties {
    /**
     * @description id
     */
    id: number;

    /**
     * @description name
     */
    name: string;

    /**
     * @description date
     */
    date: Date;
}

export interface IResponse {
    unsortableProperty: string;
}

export interface IRequest {
    sortBy: Sortify<Pick<ISortableProperties, "id">> | Sortify<Pick<ISortableProperties, "name">> | Sortify<Pick<ISortableProperties, "date">>
}

export type Sortify<T> = {
    [P in keyof T]: sorting.SortDirection
}
```

The schema I'd like to see from this is 
```
{
  "$schema": "http://json-schema.org/draft-06/schema#",
  "additionalProperties": false,
  "definitions": {
    "sorting.SortDirection": {
      "enum": [
        "ASC",
        "DESC"
      ],
      "type": "string"
    }
  },
  "properties": {
    "sortBy": {
      "anyOf": [
        {
          "additionalProperties": false,
          "properties": {
            "id": {
              "$ref": "#/definitions/sorting.SortDirection",
              "description": "id"
            }
          },
          "required": [
            "id"
          ],
          "type": "object"
        },
        {
          "additionalProperties": false,
          "properties": {
            "name": {
              "$ref": "#/definitions/sorting.SortDirection",
              "description": "name"              
            }
          },
          "required": [
            "name"
          ],
          "type": "object"
        },
        {
          "additionalProperties": false,
          "properties": {
            "date": {
              "$ref": "#/definitions/sorting.SortDirection",
              "description": "date"              
            }
          },
          "required": [
            "date"
          ],
          "type": "object"
        }
      ]
    }
  },
  "required": [
    "sortBy"
  ],
  "type": "object"
}

```

But I actually get the error :
ts-json-schema-generator\dist\ts-json-schema-generator.js:30
        throw error;
        ^

TypeError: Cannot read property 'reduce' of undefined.

Is there any possibility of supporting this kind of scenario?

Thanks!
Rob

**Comment by domoritz** [comment]
Simple example

```ts
export type MyObject = Part<B>;

type Part<T> = {
    [P in keyof T]?: T[P];
};

interface B {
    b: number;
    c: string;
}
```

The reason why it doesn't work is that we need to associate keys with values in `Part`. I didn't find an obvious way to do that and don't have time right now but will look at it in a few weeks.

**Comment by domoritz** [comment]
@robertdempsey it shouldn't take too long to fix this and I would love a PR for this. We could use this to fix an issue here: https://github.com/vega/vega-lite/pull/3491#discussion_r173636920

**Comment by robertdempsey** [comment]
@domoritz sorry I haven't replied yet. We are a bit stretched for time too but will see if we can take a look. It might take a couple weeks if we do though and it'd be our first time working on this library. Will keep you updated on whether or not we get to take a look. Cheers!

**Comment by renewooller** [comment]
Hi, I work with @robertdempsey. I've looked into this a little - it seems that because of the intentional limitation with the TS compiler to not know what is implementing a generic, we're going to have a hard time getting the information we need and fitting it into the way this project is set up. 
eg  see [https://github.com/Microsoft/TypeScript/issues/7391]

**Comment by domoritz** [comment]
I looked a bit into it yesterday and I can’t say for sure whether this will work but my idea was to figure out all the possible values for P and then getting the type of each property of T. This way we should get Pick and Partial working. Sortify from your example at the top should be easy because the type on the right side does not depend on a key.

**Comment by domoritz** [comment]
https://astexplorer.net/ is super helpful to see what is different.

**Comment by renewooller** [comment]
I did had a look at that - still have a bit to learn to make it useful I think. Have you ever considered using [simple-ast](https://github.com/dsherret/ts-simple-ast)?

**Comment by renewooller** [comment]
Just putting notes here to help with dev - no need to respond unless you want to.

So the point at which we're losing the information is in the `ExpressionWithTypeArgumentsNodeParser`, when we call `this.childNodeParser.createType`, the node is passes through is exactly 

```type Partial<T> = {
    [P in keyof T]?: T[P];
};```

and this has none of the type information that would allow the `baseType` to have properties.

**Comment by domoritz** [comment]
> Have you ever considered using simple-ast?

No, I didn't know about it. Looks neat.

**Comment by renewooller** [comment]
In the `MappedTypeGenerator` the `ObjectProperties` can be extracted from the arguments like so:
```
context.getArguments().reduce((acc : ObjectProperty[], arg : any) => {
                acc = acc.concat(arg.getProperties())
                return acc;
            }, [])
```
however, they are still 'required'. I'm currently trying to figure out a nice way of ensuring the attributes of the parameters defined by the mapped type are used.

**Comment by renewooller** [comment]
Well, I've got an ugly implementation working. Will do a pull request.

**Comment by domoritz** [comment]
Awesome. I'll review it once it's cleaned up.

**Comment by renewooller** [comment]
That's all linted.. Looking at Pick now.

**Comment by domoritz** [comment]
Fantastic. I'm really excited to have your contribution. There are a couple of pending refactorings that we couldn't do because of missing support for Partial. For instance https://github.com/vega/vega-lite/pull/3502.

**Comment by domoritz** [comment]
Fixed in #ebad910
